onnx/cpp2py_export.cc

```diff
@@ -139,63 +139,63 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
+        ParseProtoFromPyBytes(proto, bytes);
+        checker::check_value_info(proto, ctx);
       });
 
   checker.def(
       "check_tensor",
       [](const py::bytes& bytes, const checker::CheckerContext& ctx) -> void {
-        std::unique_ptr<TensorProto> proto(new TensorProto());
-        ParseProtoFromPyBytes(proto.get(), bytes);
-        checker::check_tensor(*proto, ctx);
+        TensorProto proto{};
+        ParseProtoFromPyBytes(proto, bytes);
+        checker::check_tensor(proto, ctx);
       });
 
   checker.def(
       "check_attribute",
       [](const py::bytes& bytes, const checker::CheckerContext& ctx) -> void {
-        std::unique_ptr<AttributeProto> proto(new AttributeProto());
-        ParseProtoFromPyBytes(proto.get(), bytes);
-        checker::check_attribute(*proto, ctx, checker::LexicalScopeContext());
+        AttributeProto proto{};
+        ParseProtoFromPyBytes(proto, bytes);
+        checker::check_attribute(proto, ctx, checker::LexicalScopeContext());
       });
 
   checker.def(
       "check_node",
       [](const py::bytes& bytes, const checker::CheckerContext& ctx) -> void {
-        std::unique_ptr<NodeProto> proto(new NodeProto());
-        ParseProtoFromPyBytes(proto.get(), bytes);
+        NodeProto proto{};
+        ParseProtoFromPyBytes(proto, bytes);
         checker::LexicalScopeContext lex_ctx;
-        checker::check_node(*proto, ctx, lex_ctx);
+        checker::check_node(proto, ctx, lex_ctx);
       });
 
   checker.def(
       "check_graph",
       [](const py::bytes& bytes, const checker::CheckerContext& ctx) -> void {
-        std::unique_ptr<GraphProto> proto(new GraphProto());
-        ParseProtoFromPyBytes(proto.get(), bytes);
+        GraphProto proto{};
+        ParseProtoFromPyBytes(proto, bytes);
         checker::LexicalScopeContext lex_ctx;
-        checker::check_graph(*proto, ctx, lex_ctx);
+        checker::check_graph(proto, ctx, lex_ctx);
       });
 
   checker.def("check_model", [](const py::bytes& bytes) -> void {
-    std::unique_ptr<ModelProto> proto(new ModelProto());
-    ParseProtoFromPyBytes(proto.get(), bytes);
-    checker::check_model(*proto);
+    ModelProto proto{};
+    ParseProtoFromPyBytes(proto, bytes);
+    checker::check_model(proto);
   });
 
   // Submodule `optimizer`
   auto optimizer = onnx_cpp2py_export.def_submodule("optimizer");
   optimizer.doc() = "Optimizer submodule";
 
   optimizer.def(
-      "optimize", [](const py::bytes& bytes, std::vector<std::string>& names) {
-        std::unique_ptr<ModelProto> proto(new ModelProto());
-        ParseProtoFromPyBytes(proto.get(), bytes);
-        std::unique_ptr<ModelProto> result(
-            optimization::Optimize(std::move(proto), names));
+      "optimize",
+      [](const py::bytes& bytes, const std::vector<std::string>& names) {
+        ModelProto proto{};
+        ParseProtoFromPyBytes(proto, bytes);
+        auto const result = optimization::Optimize(std::move(proto), names);
```

Comment:
In the original implementation, we only pass one unique_ptr, so the caller (`cpp2py`) and the callee (`Optimize`) share the same instance. And in `Optimize` (after the std::move(proto)), the callee own the `proto`. When `Optimize` finish, the proto will be released.

In the proposed one, we pretty much do the same thing. The only difference is we transfer the ownership to the parameter of `Optimize`. So the first parameter of `Optimize` owns the proto. So the life scope of the proto ends at the end of `Optimize` function.

So I don't see big win here.

Reply:
"In the proposed one, we pretty much do the same thing. The only difference is we transfer the ownership to the parameter of Optimize."
No, after this change Optimize doesn't have ownership on the ModelProto.

Note in the case you have a choice to either create an object or unique_ptr (like here in the pybind11 wrapper), memory usage won't make too much difference (although I can still argue it's more preferable to allocate on stack rather than doing unnecessary dynamic allocation). But as I have explained in the discussion in "optimize.h", in case the caller gets ModelProto from somewhere else, using the unique_ptr in the optimizer interface would force the caller to make another copy (and causes more memory usage).

But more importantly, the previous interface really confuses users: why does the optimizer need to own a copy of the ModelProto (does it imply it will actually do in-place modification?)

Reply:
agree on this point "But more importantly, the previous interface really confuses users: why does the optimizer need to own a copy of the ModelProto (does it imply it will actually do in-place modification?)"

I'm going to propose that we should do in-place modification to avoid copy (especially now we're copying all weights into graph, which should be avoided, I think).

Reply:
@linkerzhang I don't quite understand which copy are you trying to avoid:

There are four phases:
1. caller pass ModelProto to optimizer
2. optimizer converts ModelProto to Graph
3. optimizer passes Graph to each pass
4. optimizer converts Graph back to ModelProto
5. optimizer returns the new ModelProto

1 won't be doing copy anymore after this change, 3 is doing in-place modification already so no copy, 2 and 4 are by design doing copy so passes writer can more easily manipulate the underlying data (and to be fair not very related to this PR). 5 doesn't have copy thanks to RVO (and thanks to @yinghai pointed this out then leads to this PR).

Reply:
I still cannot see how the original implementation will create more copies of ModelProto than the new one at step 1. Could you explain more?

Reply:
Assuming we have protobuf <3.4.0 and we don't have move ctor for Message, it really depends on what caller of optimizer is using. If it uses 
auto optimized_model = std::make_unique<ModelProto>(optimize(old_model, {}));
Current PR will incur an copy. 
On the other hand, if the caller uses plain ModelProto and we have the old interface, it will need to copy the ModelProto in the unique_ptr, too. 

Either way it's not perfect. The interface in this PR is more natural IMO. If you limit your interface to `unique_ptr`, calls to this function will contain quite a bit of `std::move`.

Reply:
I meant the copy in 2 could be avoided with unique_ptr interface, but can't be avoided with const reference interface. @bddppq